broadcast/src/main/java/org/elasticsoftware/elasticactors/broadcast/handlers/RehashHandlers.java

```diff
@@ -0,0 +1,110 @@
+                actorRef.tell(new RehashRequestInternal());
+            }
+        }
+    }
+
+    @MessageHandler
+    public void handle(RehashRequestInternal rehashRequest, BroadcasterState state, ActorRef sender) {
+        if (state.getCurrentlyRehashing()) {
+            logger.warn(String.format("Broadcaster actor <%s> received rehash request, but it only contains a leaf node. Ignoring.", getSelf().getActorId()));
+            return;
+        }
+
+        if (state.isLeafNode()) {
+            sender.tell(new RehashResponse(state.getLeaves()));
+        } else {
+            state.setCurrentlyRehashing(true);
+            state.setRehashReplyTo(sender);
+            state.setRehashMembers(new HashSet<ActorRef>());
+            state.setExpectedRehashingReplies(state.getNodes().size());
+            state.setReceivedRehashingReplies(0);
+
+            for (ActorRef actorRef : state.getNodes()) {
+                actorRef.tell(new RehashRequestInternal());
+            }
+        }
+    }
+
+    @MessageHandler
+    public void handle(RehashResponse rehashResponse, BroadcasterState state) throws Exception {
+        state.incrementReceivedRehashingReplies();
+        state.getRehashMembers().addAll(rehashResponse.getMembers());
+
+        if (state.getReceivedRehashingReplies().equals(state.getExpectedRehashingReplies())) {
+            if (state.getRehashRoot()) {
+                // this node is the root of the broadcaster, once all replies are received
+                // the tree is deleted & recreated
+                for (ActorRef actorRef : state.getNodes()) {
+                    getSystem().stop(actorRef);
+                }
+
+                state.setLeafNode(true);
+                state.getLeaves().clear();
+                state.getNodes().clear();
+
+                logger.info(String.format("Rehashing of broadcaster <%s> is now completed, re-adding all members to the tree", getSelf().getActorId()));
+
+                getSelf().tell(new Add(state.getRehashMembers()));
+
+                if (state.getRehashReplyTo() != null) {
+                    state.getRehashReplyTo().tell(new RehashComplete());
+                }
+            } else {
+                // this node is not the root of the broadcaster, send all it's children further up the tree
+                state.getRehashReplyTo().tell(new RehashResponse(state.getRehashMembers()));
+            }
+
+            // this node has received all replies it was waiting for, time to mark the state as such
+            state.setCurrentlyRehashing(false);
+            state.setRehashReplyTo(null);
+            state.setRehashMembers(new HashSet<ActorRef>());
+            state.setExpectedRehashingReplies(0);
+            state.setReceivedRehashingReplies(0);
+        }
+    }
+
+    @Message(serializationFramework = JacksonSerializationFramework.class, durable = true, immutable = true)
+    private static final class RehashRequestInternal {
```

Comment:
Hi Bogdan, I think it would be best to have this class at the highest level as I am not 100% sure that the MessagesScanner will correctly identity it